scripts/hail_batch/project_wgs_onto_snp_chip_pca/project_wgs_samples_onto_snp_chip.py

```diff
@@ -0,0 +1,113 @@
+"""
+Project WGS data onto SNP-chip data
+"""
+
+import re
+import hail as hl
+from analysis_runner import bucket_path, output_path
+from hail.experimental import pc_project
+from hail.experimental import lgt_to_gt
+from bokeh.plotting import ColumnDataSource, figure
+from bokeh.palettes import Dark2  # pylint: disable=no-name-in-module
+from bokeh.transform import factor_cmap
+from bokeh.resources import CDN
+from bokeh.embed import file_html
+from bokeh.io.export import get_screenshot_as_png
+
+SNP_CHIP = bucket_path('snpchip/v1/snpchip_grch38.mt')
+TOB_WGS = bucket_path('mt/v3-raw.mt')
+
+
+def query():
+    """Query script entry point."""
+
+    hl.init(default_reference='GRCh38')
+
+    snp_chip = hl.read_matrix_table(SNP_CHIP).key_rows_by('locus', 'alleles')
+    tob_wgs = hl.read_matrix_table(TOB_WGS)
+    tob_wgs = hl.experimental.densify(tob_wgs)
+    tob_wgs = tob_wgs.annotate_entries(GT=lgt_to_gt(tob_wgs.LGT, tob_wgs.LA))
+
+    snp_chip = snp_chip.semi_join_rows(tob_wgs.rows())
+    # Perform PCA
+    eigenvalues, scores, loadings = hl.hwe_normalized_pca(
+        snp_chip.GT, compute_loadings=True, k=5
+    )
+
+    # make tob_wgs rows equivalent to the snp_chip rows
+    tob_wgs = tob_wgs.semi_join_rows(snp_chip.rows())
+    snp_chip = snp_chip.annotate_rows(af=hl.agg.mean(snp_chip.GT.n_alt_alleles()) / 2)
+    loadings = loadings.annotate(af=snp_chip.rows()[loadings.key].af)
+    # project WGS samples onto PCA
+    ht = pc_project(tob_wgs.GT, loadings.loadings, loadings.af)
+    scores = scores.key_by(s=scores.s + '_snp_chip')
+    union_scores = ht.union(scores)
+    variance = [(x / sum(eigenvalues) * 100) for x in eigenvalues]
+    variance = [round(x, 2) for x in variance]
+
+    # Get partner sample information
+    sample_names = union_scores.s.collect()
+
+    def sample_type(sample_name):
+        if sample_name.endswith('snp_chip'):
+            partner_name = re.sub('_snp_chip', '', sample_name)
+            tech = 'snp'
+        else:
+            partner_name = sample_name + '_snp_chip'
+            tech = 'wgs'
+
+        if partner_name in sample_names:
+            prefix = 'dual_'
+        else:
+            prefix = ''
+
+        return prefix + tech
+
+    # plot
+    labels = list(map(sample_type, sample_names))
+    cohort_sample_codes = list(set(labels))
+    tooltips = [('labels', '@label'), ('samples', '@samples')]
+
+    # Get number of PCs
+    number_of_pcs = len(eigenvalues)
+
+    for i in range(0, (number_of_pcs - 1)):
+        pc1 = i
+        pc2 = i + 1
+        plot = figure(
+            title='TOB-WGS + TOB SNP Chip',
+            x_axis_label='PC' + str(pc1 + 1) + ' (' + str(variance[pc1]) + '%)',
+            y_axis_label='PC' + str(pc2 + 1) + ' (' + str(variance[pc1]) + '%)',
```

Comment:
Do you want to use the f-strings from the previous plot? That'll also fix the legend (`pc1` instead of `pc2`).

Reply:
Good idea, I've now updated this to use f-strings instead.